Client/game_sa/CRenderWareSA.cpp

```diff
@@ -496,30 +497,70 @@ bool AtomicsReplacer(RpAtomic* pAtomic, void* data)
+    if (RpClumpGetNumAtomics(pSrc) > 1 || (RpClumpGetNumAtomics(pSrc) == 1 && pModelInfo->GetModelType() == eModelInfoType::CLUMP))
     {
-        RpAtomic* pOldAtomic = (RpAtomic*)pModelInfo->GetRwObject();
+        // Get new interface but with old RwObject
+        auto* currentNewInterface = static_cast<CClumpModelInfoSAInterface*>(pModelInfo->GetInterface());
+
+        // Destroy old RwObject (atomic type)
+        // We need to remove the RwObject from the original interface because
+        // the new interface points to the CClumpModelInfo vtbl
+        CBaseModelInfoSAInterface* originalInterface = pModelInfo->GetOriginalInterface();
+        if (!originalInterface)
+            return false;
+
+        if (RpClumpGetNumAtomics(reinterpret_cast<RpClump*>(originalInterface->pRwObject)) == 1)
+            reinterpret_cast<CAtomicModelInfoSAInterface*>(originalInterface)->DeleteRwObject();
+        else
+            currentNewInterface->DeleteRwObject();
+
+        currentNewInterface->pRwObject = nullptr;
 
-        if (reinterpret_cast<RpClump*>(pOldAtomic) != pNew && !DoContainTheSameGeometry(pNew, NULL, pOldAtomic))
+        // Init new RwObject (clump type)
+        currentNewInterface->SetClump(pSrc);
+    }
+    else
+    {
+        auto* currentInterface = static_cast<CAtomicModelInfoSAInterface*>(pModelInfo->GetInterface());
+
+        // Destroy old RwObject (clump type)
+        // We need to remove the RwObject from the original interface because
+        // the new interface points to the CAtomicModelInfo vtbl
+        CBaseModelInfoSAInterface* originalInterface = pModelInfo->GetOriginalInterface();
+        if (originalInterface && RpClumpGetNumAtomics(reinterpret_cast<RpClump*>(originalInterface->pRwObject)) > 1)
         {
-            // Clone the clump that's to be replaced (FUNC_AtomicsReplacer removes the atomics from the source clump)
-            RpClump* pCopy = RpClumpClone(pNew);
+            reinterpret_cast<CClumpModelInfoSAInterface*>(originalInterface)->DeleteRwObject();
+            currentInterface->pRwObject = nullptr;
+        }
+        else
+            currentInterface->DeleteRwObject();
 
-            // Replace the atomics
-            SAtomicsReplacer data;
-            data.usTxdID = ((CBaseModelInfoSAInterface**)ARRAY_ModelInfo)[usModelID]->usTextureDictionary;
-            data.pClump = pCopy;
+        RpAtomic* atomic = GetFirstAtomic(pSrc);
+        RwFrame*  frame = RpGetFrame(atomic);
 
-            MemPutFast<DWORD>((DWORD*)DWORD_AtomicsReplacerModelID, usModelID);
-            RpClumpForAllAtomics(pCopy, AtomicsReplacer, &data);
+        char name[24];
+        bool damage = false;
+        GetNameAndDamage(GetFrameNodeName(frame), name, damage);
 
-            // Get rid of the now empty copied clump
-            RpClumpDestroy(pCopy);
-        }
+        if (damage)
+            static_cast<CDamageableModelInfoSAInterface*>(currentInterface)->SetDamagedAtomic(atomic);
```

Comment:
OOB and causes random crashes if the model is atomic

Reply:
I don't understand what do you mean. CDamageableAtomic inherit CAtomicModelInfo

Reply:
And has an extra field. You can't safely cast CAtomicModelInfo to CDamageableAtomic and use that field.
You write out of buffer here and override random data. This causes random crashes.
This is why 'object-damageable' exists in engineRequestModel

Reply:
So I think it would be safer to simply leave SetAtomic without considering SetDamagedAtomic because it's better than checking the vtbl, right?

Reply:
Check if it works.